File: src/org/jetbrains/haskell/sdk/HaskellSdkAdditionalData.java
package org.jetbrains.haskell.sdk;

import com.intellij.openapi.options.ConfigurationException;
import com.intellij.openapi.projectRoots.SdkAdditionalData;
import com.intellij.openapi.projectRoots.SdkModel;
import org.jdom.Element;
import org.jetbrains.annotations.NotNull;

public final class HaskellSdkAdditionalData implements SdkAdditionalData {
    private static final String GHC_OPTIONS = "GhcOptions";
    private static final String CABAL_PATH = "CabalPath";

    private String myGhcOptions = "";
    private String myCabalPath = "";

    public HaskellSdkAdditionalData() {

    }

    public HaskellSdkAdditionalData(Element element) {
        this.myGhcOptions = element.getAttributeValue(GHC_OPTIONS);
    }

    public void checkValid(SdkModel sdkModel) throws ConfigurationException {
    }

    @Override
    public Object clone() throws CloneNotSupportedException {
        return super.clone();
    }

    public void save(Element element) {
        if (myGhcOptions != null) {
            element.setAttribute(GHC_OPTIONS, myGhcOptions);
            element.setAttribute(CABAL_PATH, myCabalPath);
        }
    }

    public String getGhcOptions() {
        return myGhcOptions == null ? "" : myGhcOptions;
    }

    public void setGhcOptions(String ghcOptions) {
        this.myGhcOptions = ghcOptions;
    }

    @NotNull
    public String getCabalPath() {
        return myCabalPath == null ? "" : myCabalPath;
    }

    public void setCabalPath(String cabalPath) {
        this.myCabalPath = cabalPath;
    }
}

File: src/org/jetbrains/haskell/sdk/HaskellSdkAdditionalData.kt
package org.jetbrains.haskell.sdk

import com.intellij.openapi.options.ConfigurationException
import com.intellij.openapi.projectRoots.SdkAdditionalData
import com.intellij.openapi.projectRoots.SdkModel
import org.jdom.Element

private val GHC_OPTIONS = "GhcOptions"
private val CABAL_PATH = "CabalPath"

public class HaskellSdkAdditionalData(ghcOptions: String?, cabalPath: String?) : SdkAdditionalData, Cloneable {
    private var myGhcOptions: String? = ghcOptions
    private var myCabalPath: String? = cabalPath

    public fun checkValid(sdkModel: SdkModel?): Unit {

    }

    public override fun clone() : Any {
        return super<Cloneable>.clone()
    }
    public fun save(element: Element): Unit {
        if (myGhcOptions != null)
        {
            element.setAttribute(GHC_OPTIONS, myGhcOptions)
            element.setAttribute(CABAL_PATH, myCabalPath)
        }

    }
    public fun getGhcOptions(): String? {
        return (if (myGhcOptions == null)
            ""
        else
            myGhcOptions)
    }
    public fun setGhcOptions(ghcOptions: String?): Unit {
        this.myGhcOptions = ghcOptions
    }
    public fun getCabalPath(): String {
        return if (myCabalPath == null) "" else myCabalPath!!
    }
    public fun setCabalPath(cabalPath: String?): Unit {
        this.myCabalPath = cabalPath
    }


    class object {
        public fun init(): HaskellSdkAdditionalData {
            return HaskellSdkAdditionalData(null, null)
        }
        public fun load(element: Element): HaskellSdkAdditionalData {
            val data = HaskellSdkAdditionalData(null, null)
            data.myCabalPath = element.getAttributeValue(CABAL_PATH)
            data.myGhcOptions = element.getAttributeValue(GHC_OPTIONS)
            return data
        }

    }
}
